### ISSUE TITLE
Architecture: simplify dual-API hash dispatch

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#6471]
Currently (as of the upcoming Mbed TLS 3.3), there are two ways that code can call hash functions: via the MD module, or via PSA. We handle that with an extra abstraction layer implemented in `hash_info.{h,c}` for metadata (algorithm availability and representation, etc.), and code that either calls `mbedtls_md()` or `psa_hash_compute()` (or their variants) depending on whether `MBEDTLS_MD_C` is enabled.

We could simplify this code and hide the complexity inside `md.c`. I'm not convinced that there's a good reason not to. If a hash algorithm is available through PSA, as determined by compile-time options, then `md.c` can dispatch through PSA. The caller doesn't need to know.

This can only work if the compile-time options are enough to guarantee that the PSA code path works. For most cryptographic operations, this is not the case because they require the PSA keystore, the PSA RNG, or both, and those are not available until `psa_crypto_init` has been called. However, hash operations don't require the library to have been initialized — this has always been the case, but [was not previously documented or tested](https://github.com/Mbed-TLS/mbedtls/pull/6470). There is one limitation, which is a [bug](https://github.com/Mbed-TLS/mbedtls/issues/6469) we may need to solve first: `psa_crypto_init` is necessary if the hash is accelerated and the hash driver has an initialization function.

The main reason for this simplification is the code complexity. I expect it to be neutral in terms of code size. The complexity:

* has led us to introduce an extra abstraction layer (`hash_info.h`) which is largely redundant with `md.h`.
* is requiring us to do extra work in each library module that uses a hash. This work has only partially been done so far.
* is causing extra mistakes and difficulties, e.g. https://github.com/Mbed-TLS/mbedtls/pull/6451.

The goal of this task is to work out the architecture and figure out what we'd need to do, and whether it's worth it at this stage (considering that the implementation of `hash_info` is already partially underway). Would we be redoing work that's already good enough, or would we stop throwing good money after bad?

----

There is a subtlety: if you build Mbed TLS as a TLS library where the PSA crypto layer is only an RPC client and all the PSA cryptography is happening in a remote service (`MBEDTLS_PSA_CRYPTO_CLIENT` without `MBEDTLS_PSA_CRYPTO_C`), but some cryptography is also available through the legacy API. Suppose the client code (which is third-party code, we don't have that code in Mbed TLS) requires a call to `psa_crypto_init` before any RPC call works (this is a natural architecture).

Now suppose `psa_crypto_init` hasn't been called yet. Then `mbedtls_md(MBEDTLS_MD_SHA256)` works and invoke the code in `sha256.c`, whereas `psa_hash_compute(PSA_ALG_SHA_256)` would need to make an RPC call, which fails if `psa_crypto_init` hasn't been called. So we can't indiscriminately dispatch md calls to PSA.

Currently, this subtlety is moot because we don't allow `MBEDTLS_USE_PSA_CRYPTO` with `MBEDTLS_PSA_CRYPTO_CLIENT`.

----

**Comment by mpg** [comment]
> considering that the implementation of `hash_info` is already partially underway

There is a misunderstanding: `hash_info` is expected to stay as it is (until it's eventually removed when we move be purely based on PSA Crypto). What makes you think it's "partially" underway?

> Currently, this subtlety is moot because we don't allow `MBEDTLS_USE_PSA_CRYPTO` with `MBEDTLS_PSA_CRYPTO_CLIENT`.

`MBEDTLS_USE_PSA_CRYPTO` is only about TLS, X.509 and PK. Other modules, such as RSA, PKCS5, PKCS12, are not governed by this option, and those are the places that currently use `hash_info` and the pattern `#if defined(MD_C) /* call MD */ #else /* call PSA */ #endif` that you're unhappy about, so I'm not quite sure why you're mentioning `MBEDTLS_USE_PSA_CRYPTO` here.

**Comment by mpg** [comment]
To clarify regarding `hash_info`: currently MD is used for two things:

1. managing information about hashes, mostly output size (and block size);
2. actually doing hash operations.

`hash_info`, as the name implies, is meant to be only about information, so it's not expected to grow.

**Comment by mpg** [comment]
> > considering that the implementation of `hash_info` is already partially underway
> 
> There is a misunderstanding: `hash_info` is expected to stay as it is

Or when you write `hash_info` you're not specifically talking about the files `hash_info.[ch]` but rather you're using that as a shortcut name for the general architecture? I think most of your issues are rather with the `#if defined(MD_C) /* call MD */ #else /* call PSA */ #endif` (sometimes #if defined(MD5_C) /* call md5 */ #else /* call PSA */ #endif`) pattern and the admittedly complicated set of `MBEDTLS_HAVE_ALG_xxx_VIA_yyy` macros that go with it.

So far I think the design you're suggestion will not allow as much code savings as the current design in the most favorable case (all hashes via drivers), but I agree that a major downside of the current design is complexity.

**Comment by mpg** [comment]
Just dumping a thought I have about it over the week-end: there's one module (PEM) which is doing hash computations not by calling MD, but directly `mbedtls_md5` functions. If we were to make MD the single abstraction, this means we need to use `mbedtls_md()` functions instead, which is not a hard change to do, but this means PEM (or more specifically, this optional feature of PEM) now depends on `MD_C`. In the past this would have been a compatibility issue, but fortunately we now have the ability of auto-enabling `MD_C` when both PEM and MD5 are available, so this is something we could do if decided to make MD the single abstraction for hashes.

**Comment by ronald-cron-arm** [comment]
I am a bit surprised by this issue. My understanding is that eventually we want to get rid of md.c and provide hash operations only through the PSA API. I understand that the current situation is probably not ideal as hash information in some contexts (like in SSL ciphersuite) is still the md way (and we need to translate it) but we stop calling md.c and rather call the PSA interface which is where we want to go eventually if I understand correctly. Moving back to using the md.c interface does not seem to go in the right direction.

**Comment by gilles-peskine-arm** [comment]
> there's one module (PEM) which is doing hash computations not by calling MD, but directly mbedtls_md5 functions

It's not the only one. There's also RSA (only in the self-test code), X.509, entropy and TLS.

**Comment by gilles-peskine-arm** [comment]
> we stop calling md.c and rather call the PSA interface

That's not what we're doing. We're now calling yet another abstraction (`hash_info`) for some purposes, and duplicating the code for the parts that `hash_info` doesn't implement.

**Comment by gilles-peskine-arm** [comment]
If the objection is just “we've started on hash_info and we don't reconsider past decisions”, I got nothing.

If the objection is “md is old and we want to get rid of it”, I don't see the urgency. We aren't ready to get rid of it yet, so why force it? But we could actually completely switch to PSA as the interface for hashes in our code, with a thin wrapper the other way for builds without PSA — `psa_hash_xxx` calling `mbedtls_md_xxx` (maybe even through macros? I'm not sure that's possible). But that doesn't solve the problem of third-party code using MD, and library code using MD for HMAC, not using PSA hash accelerators.

**Comment by gilles-peskine-arm** [comment]
> There is a subtlety: if you build Mbed TLS as a TLS library where the PSA crypto layer is only an RPC client and all the PSA cryptography is happening in a remote service (`MBEDTLS_PSA_CRYPTO_CLIENT` without `MBEDTLS_PSA_CRYPTO_C`)

That's incompatible with `MBEDTLS_USE_PSA_CRYPTO`, but we're expanding the use of PSA interfaces to builds without `MBEDTLS_USE_PSA_CRYPTO`, so this is a real problem.

**Comment by gilles-peskine-arm** [comment]
Note: in this and related discussions, I've been discussing “hash_info”, but what I really meant was the combination of three things: the macros defined in `include/mbedtls/legacy_or_psa.h`, the duplication of code paths in each module using md for hashes, and the partial abstraction `hash_info.h`.

**Comment by gilles-peskine-arm** [comment]
### Potential problem: PSA can't do everything legacy can do

It's currently possible, and not completely unreasonable, to have a build with e.g. `MBEDTLS_PSA_CRYPTO_C`, `MBEDTLS_PSA_CRYPTO_CONFIG` and `MBEDTLS_MD5_C` but not `PSA_WANT_ALG_MD5`, because the application uses MD5 for some legacy purpose (e.g. PKCS#5) but does not want to have support for MD5 in e.g. RSA-PSS signatures which are done by a driver.

So we can't just use `PSA_WANT_ALG_MD5` (which we'd define even if PSA is disabled) to gate MD5 hash calculations via md, because in this configuration, md can currently do MD5 but PSA can't.

Note that this configuration will not be possible once PSA becomes the sole crypto engine. The general JSON-based PSA crypto configuration mechanism can express things like “HMAC-MD5 but not RSA-PSS-MD5”, but the C-macro-based mechanism `PSA_WANT_`) cannot.

### Potential problems: PSA client with some local crypto

It's currently possible, and not completely unreasonable, to have a build with `MBEDTLS_PSA_CRYPTO_CLIENT` and some local cryptography using legacy APIs. This causes PSA crypto API calls to be routed to a client-server mechanism, with the cryptography performed in the server. This configuration is desirable, for example, if the server has access to cryptographic accelerators but the client doesn't, but the client needs to perform some cryptography locally before the server starts (typically, in a non-secure boot loader which will eventually load a signed image for the secure world, with a crypto server in the secure world).

In this configuration, API function names like `psa_hash_compute` are not provided by Mbed TLS. They are provided by the client side of the third-party client-server software. We must not define a function or macro with this name. (A macro could be acceptable for internal use, but it would be incompatible with `#include <psa/crypto.h>`.)

Note that this configuration will not be possible once PSA becomes the sole crypto engine.

**Comment by ronald-cron-arm** [comment]
First, to be more specific, what I had in mind when I wrote my comment above was: in TLS code we have some calls to psa_hash_xxx() functions (under some well chosen pre-processor directives) and I thought it would not go in the right direction to move that back to call to  mbedtls_md_... (but I may be wrong on that). I thought that issue was suggesting to do that thus my comment.

But looking at #6474, I do not see any change regarding the calls to psa_hash_xxx() in ssl_tls.c thus either I did not understand right or this is not done yet or some things are not aligned with this issue and the PR #6474 ...

Otherwise there are a lot of things going on here and I am not sure to be able to do pertinent and useful comments. Trying to participate to the debate though:
1) I agree that legacy_or_psa.h being a public header does not sound good. Not possible to keep it internal ?
2) Dispatching via PSA in the MD module seems to go in the right direction to me as I think we had in mind for MD to be a wrapper around PSA  at some point. But even if we have that I would prefer to keep in TLS (talking about the code I know here) the direct calls to PSA when they are already in place.
3) Regarding hash_info , md and code duplication, wouldn't it possible to reduce the code size by rather disabling some code in md : keeping only the relevant part when we have PSA hashing + USE_PSA_CRYTO ?

> If the objection is just “we've started on hash_info and we don't reconsider past decisions”, I got nothing.
Not at all what I have/had in mind.

**Comment by mpg** [comment]
One thing one might dislike with the current design is that in builds with MD included (which are most builds except for people who have drivers for all the hashes they need and try hard to save as much code as possible) they're some amount of code/data duplication between `hash_info.c` and the parts of MD that do similar things. Let me explain why I think this is a pretty minor issue.

First, `hash_info.o` is 236 bytes, not all of which is duplicating things from MD, because it includes conversion between MD and PSA identifiers, which MD does not (see below). As comparison point, I'll note that [in a recent PR](https://github.com/Mbed-TLS/mbedtls/pull/6282#discussion_r985485915) we decided 280 bytes for a new feature was not enough to justify adding a new compile-time option for that feature.

Second, the introduction of `hash_info.c` unified conversion between MD and PSA identifiers, which was previously using a `static inline` function included in several files. I believed this resulted in a strict improvement over the existing, based on the following numbers (total size of the 3 libraries with `-mcpu=cortex-m0 -mthumb -Os`):

| commit | description | size in bytes |
| --- | --- | --- |
| 9115e1bc4d00 | before #6065 | 414402 |
| 1f7f7172dcaa | right before `hash_info.c` | 405516 |
| d82a9edc636b | `hash_info.c` present and used | 405089 |
| 68429fc44da6 | end of #6065 | 405089 |

Third, if we really care, I'm sure we can still find a way to share more data and reduce duplication between `hash_info.c` and `md.c`, though to me honest I'm not sure why this particular 200-bytes inefficiency would be more urgent to resolve that other, possibly larger, inefficiencies I'm sure we have elsewhere.

A minor vocabulary point: I'm not sure I agree with calling this yet another abstraction layer, as it's not implemented on top of other layers, I rather see it as an independent module for retrieving information.

(For the record, I take the point that there are undocumented and untested properties of the functions in `hash_info.h` that the rest of the code relies on, which should and will be fixed if we keep this module.)

**Comment by gilles-peskine-arm** [comment]
> Practical: centralizing to MD means more work now and in the future: now obviously because it would be a significant change, so more cost than not doing any change; and in the future because if we remove the code path using psa_hash now, we'll have to write it again in the future - which is not hard in itself, but tedious as there are many places. So, one question is: is that cost worth it?

Doing a systematic change is easier than doing ad hoc changes. But it's true that it's possible that the remaining ad hoc change will be just eliminating bits of the code. Not just eliminating the `#else` parts but also things like error code conversions.

> Understanding: a long time ago (but in a galaxy not far away at all), for code obeying USE_PSA_CRYPTO, we decided that we would not use MD but instead have two code paths, one of them calling psa_hash directly, and that the duplication was not an issue. A bit more recently, we decided the same for cipher (and executed on it). I'm not saying we should never revisit past decisions, but I'd like to understand why. Do we think we made a mistake and this was never the right decision - if so, what did we miss? Or was it the right decision at the time but the situation is different now - if so, how?

Yes, I think that was a bad choice. It was, and still is, premature optimization. We should instead have made md.h, cipher.h and pk.h dispatch to PSA, so that as much as possible USE_PSA_CRYPTO was confined inside those modules. Not completely: setting the key needed to have different code paths, but we should have kept the same context types (with, at least for USE_PSA_CRYPTO, opaque contents that can change over time) and the same functions to use the key once set.

**Comment by mpg** [comment]
Regarding my last point, two possible differences that could be relevant.

**Dependencies.** Until now, we've been content with using `MBEDTLS_SHA256_C` as the universal way to express "Can we use SHA-256?" even when we were actually using `psa_hash` to do the computation. This works well enough except when it doesn't, that is when SHA-256 is provided only by a driver, our current focus, so now we're trying to be more precise in how we express dependencies. Perhaps this is exposing a flow that was always present in the "two code paths" design but so far hidden by the fact that we were not even trying to track dependencies precisely? That is, with two code paths everywhere, we need to know which one will be taken in each place (RSA v1.5/v2.1, PKCS5, PKCS12, PEM, PK, etc.), leading to an explosion of dependency macros to cover all cases, that are hard to defined in a meaningful way, to track which one to use for which module, etc. While if everything is centralised to MD and everyone calls MD, the dependency situation becomes much more uniform, hence manageable.

**(Lack of) explicit choice.** With past work, we had an explicit macro `USE_PSA_CRYPTO` that we used to select which of the two code paths is compiled in. With the current design of recent work, I'd tried not to introduce a similar macro, out of a general principle of minimizing compile-time options, but perhaps that was misguided in this case. This means in the current design `MD_C` now has a dual role: (1) decide if MD is enabled, and (2) decide whether `psa_hash` is used in a number of places - that, except the places where it's decided based on `USE_PSA_CRYPTO`, and the places where it's decided based on `MD5_C`, `SHA1_C` or `SHA256_C` instead. So, I think `USE_PSA_CRYPTO` achieved two things: (1) giving us a guarantee that `psa_crypto_init()` has been called (which turns out not to matter for hashes, but will matter for ciphers), and (2) being an explicit a uniform (so far) indicator of whether we use legacy or PSA (in places defined as in its scope, which is PK, X.509 and TLS). 

I'm wondering (really don't have an opinion yet) whether we should have a macro that basically says, from the users perspective, "I promise to call `psa_crypto_init()` before anything else, including legacy APIs" and form our perspective "if `PSA_CRYPTO_C` is defined then you we can actually call `psa_xxx` functions and not worry". (For hashes, this wouldn't make a difference, except with `PSA_CRYPTO_CLIENT`, but this would be useful for HMAC, cipher and the rest.) This is a question that I think is relevant regardless of the design - that is whether we keep something close to the current design or decide to centralize everything to MD/Cipher.

If we keep something close to the current design a simplification we could make (unconditionally for hashes excluding HMAC, and for the rest if we have this macro), is that instead of the current mess where "are we calling PSA" is decided in a different where everywhere (`USE_PSA_CRYPTO`, `MD_C`, `SHAxxx_C`... more in the future with Cipher), it's decided in a uniform way: if `PSA_CRYPTO_C` is present as well as the macro, then PSA is used. This would I think solve some of the issues with the current design (though probably not all).

**Comment by mpg** [comment]
Note: another problem that could potentially be solved by introducing a new macro is that it would be opt-in, perhaps also making use more free to impose further restrictions, like "everything that's available on the legacy side must also be available on the PSA side" or equivalently to document that when you enable that macro, then you can only rely on algorithms that are available on the PSA side, even in legacy API. (Example: you can only do RSA-PSS with SHA-256 if you have `PSA_WANT_ALG_SHA_256`, even when using the `mbedtls_rsa_xxx()` API.)

(In the spirit of defining some corner cases out of existence / documenting that supporting them is not a goal.)

Also, going further in the same spirit: we might want to consider doing hashes only for now, and deciding that cipher will only be done later (once we've implemented split init and partial builds for PSA Crypto / released a new major version with more documented constraints on users that is less cases for us to support), if it turns out that we can find a good enough solution for hashes but not for cipher. (Differences that could possibly lead to that result: (1) hashes don't need `psa_crypto_init()` (2) there are less API discrepancies between `md.h` and `psa_hash` than between `cipher.h` and `psa_cipher`/`psa_aead`.)

Note: even though the issue title says "hash", I'm trying to consider cipher here as well, because if we change the design for hashes, we'll probably want to do the same for ciphers if possible.

**Comment by mpg** [comment]
A problem with the current design: as a user, if I have an accelerator for SHA-256 but not for SHA-512 but I still want to be able to use both algorithms in by build, I have to leave `SHA256_C` enabled. Said otherwise, I can only remove the builtin implementation if I have drivers for all the hashes I want to use.

The source of this problem is that we use the presence or absence of MD in order to decide whether to call MD or PSA, in places like RSA-PSS. So, if one hash doesn't have a driver, then MD must be present, so all hashes must be present in software. Initially I thought "well, at least you can remove the built-in implementation in one case now (when you have drivers for everything), while previously you never could, so that's a strict improvement" but the problem is that the current design doesn't really support further improvements as it is.

One way we could improve on this while mostly keeping the current design is, in places that compute hashes with `#if MD_C
 /* use MD */ #else /* use PSA /* #endif`, to switch to `#if PSA_CRYPTO_C /* use PSA */ #else /* use MD */ #endif`. This is possible for hashes because we don't need `psa_crypto_init()`. (We could extend to other cases if we introduce a macro as discussed above, by doing `#if PSA_CRYPTO_C && WE_KNOW_INIT_HAS_BEEN_CALLED` use PSA else use MD.) This solves the problem in the common case where what's available on the PSA side is a superset of what's available on the MD side (which, if possible, we should define as always true aka we explicitly don't support situations where it's not true).

Of course centralizing on MD would also solve this.

**Comment by mpg** [comment]
For the benefits of people reading this other than Gilles and me, I'd like to mention an idea we discussed verbally but I think has not been written here so far: everywhere we compute a hash, just call `psa_hash_xxx()` unconditionally. If `PSA_CRYPTO_C` is defined, great. If it's not, let's have an extra (internal!) module, auto-enabled in `build-info.h`, that provides `psa_hash_xxx()` wrappers on top of MD. That way, all the places that compute a hash have a single code path instead of the current `#ifdef` dance. Possible obstructions: see [this comment](https://github.com/Mbed-TLS/mbedtls/issues/6471#issuecomment-1288918461) - I think both of them apply. This approach probably only applies to hashes (not HMAC, no cipher/AEAD).

Open question: if we centralize on something (be it MD, or `psa_hash` as above) instead of having two code paths in each place that computes a hash, do we also to that everywhere in TLS 1.2? That is, to we keep the existing `#if USE_PSA` or do we remove them? (We never promised we would _not_ use PSA when `USE_PSA` was not defined, and the only thing that kept us from doing so what the requirement for the user to call `psa_crypto_init()`, which actually is not a problem for hashes.)

**Comment by mpg** [comment]
After further thinking and discussion about this, I now think it's indeed preferable to centralize everything to MD: that is all code that computes a hash or a HMAC should just call MD, which then transparently dispatches to PSA or software implementation. Here are the two most important advantages of this solution IMO:

1. Uniformize dependencies. Once this is done, there is a single way to check if a given hash is available, as opposed to 5 now and you have to read the source code for all the stack in order to know which one is the right one - this a a maintenance cost for us, and also a usability issue as some users are exposed to this complexity as well.
2. By centralizing dispatch to a single place, we can make it smarter, for example per-hash or even at runtime according to whether the keystore has been initialized or not, which can be quite useful in solving otherwise thorny compatibility issues.

I think the first point is the most important. Basically, it's what determines if the cost of switching to the new design (which is significant) is worth it: the current design causes significant maintainability and usability issues that cannot be fully resolved by simply tuning it a bit. Those maintainability and usability issues represent a significant cost in the medium term, which is probably larger than the costs associated to switching.

**Comment by ronald-cron-arm** [comment]
> After further thinking and discussion about this, I now think it's indeed preferable to centralize everything to MD: that is all code that computes a hash or a HMAC should just call MD, which then transparently dispatches to PSA or software implementation. Here are the two most important advantages of this solution IMO:

Just to be sure, this does not apply to TLS 1.3 code where PSA in mandatory. If we do a hash calculation in TLS 1.3 only code, we call the PSA interface directly.

**Comment by gilles-peskine-arm** [comment]
> Just to be sure, this does not apply to TLS 1.3 code where PSA in mandatory. If we do a hash calculation in TLS 1.3 only code, we call the PSA interface directly.

Right. Code that's already calling PSA, which only depends on PSA, keeps calling PSA. This includes all TLS 1.3 code, and can include code that's under `MBEDTLS_USE_PSA_CRYPTO`.

For `MBEDTLS_USE_PSA_CRYPTO`, just for ease of maintenance, we might prefer to have
```
mbedtls_md_foo();
```
instead of
```
#if defined(MBEDTLS_USE_PSA_CRYPTO)
psa_hash_foo();
#else
mbedtls_md_foo();
#endif
```
if we can guarantee that the md code path will make full use of PSA, and in particular will use the PSA accelerator if available). But explicitly writing out if-use-psa-crypto-then-use-psa-crypto is still correct.

**Comment by mpg** [comment]
As already noted, one of the problems with the current design is that you have both `HAS_ALG_xxx_VIA_MD_OR_PSA` and `HAS_ALG_xxx_VIA_LOWLEVEL_OR_PSA` macros and the only way to know which one to use it read the source.

With the new design, this annoying distinction will go away because we'll change modules that currently use hash modules directly to use MD instead.

This change is mandatory with the new design, but it's a change we could also have done with the old design, and we would have ended up with fewer families of dependency macros hence fewer headaches.

More importantly, it seems to me that this is a change we could start doing right now, without waiting for the new design to be complete, as preparation work. Specifically, the modules concerned would be:

- entropy.c - currently uses `mbedtls_sha256_xxx()` or `mbedtls_sha512_xxx()`;
- pem.c - currently uses `mbedtls_md5_xxx()` optionally if available;
- rsa.c _selftest only_ (the rest uses MD) currently uses `mbedtls_sha1() - not even clear if we need to compute the hash at runtime at all, might store the hashed result in test data instead, as in #6104 for example

Note: X.509 and TLS also use `mbedtls_shax` directly in some places (`mbedtls_x509write_crt_set_key_identifier()` for X.509, running handshake hash for TLS), but since they are governed by `USE_PSA`, the situation is different there.

@gilles-peskine-arm wdyt? Should we start migrating Entropy and PEM to use MD whenever we have the time, or do you see a reason to wait for the new design to be more advanced?

**Comment by gilles-peskine-arm** [comment]
> Should we start migrating Entropy and PEM to use MD whenever we have the time

We can't _just_ migrate modules to use md because they can currently be used in builds without `MBEDTLS_MD_C`. We can force `MBEDTLS_MD_C` on in `build_info.h`. But I think before doing that, we should optimize “MD”, both in the sense of simplifying the implementation and in the sense of extracting an “MD light” (without names, without HMAC) so only MD light would be automatically enabled. (And for HMAC, optimize by using PSA's HMAC implementation.)

We can start doing this now (in fact I already did), but the new dependencies make it harder (I didn't get to a point where CI could pass, which is what led me to start [working on 3.2 instead](https://github.com/Mbed-TLS/mbedtls/pull/6474), but I didn't port any optimizations to that branch).

I think it makes more sense to isolate and optimize (in either order) md-light before we start using it as such.

**Comment by gilles-peskine-arm** [comment]
One thing I think could be done right now is melding back `hash_info.{h,c}` into `md.{h,c}`, as in:

* move `mbedtls_hash_info_md_from_psa` into `md.c`, with the proper `#ifdef` around it
* replace the rest of the `mbedtls_hash_info` functions by calls to `mbedtls_md` functions
* define the starting point of “md light” as whatever it takes to implement the hash info stuff
* make sure it's all properly documented and tested, in particular where there are assumptions about negative results (“if this function returns a nonzero value, it means the hash is supported”).

Another thing that can be done is non-behavior-changing optimizations (of code size or complexity or both), like renumbering `MBEDTLS_MD_xxx` values to follow PSA, and using PSA macros or switch-case instead of looking up inside info structures for metadata.

**Comment by mpg** [comment]
Note: another annoying thing with the current design is having to pick the right size macros under penalty of buffer overflow, see for example https://github.com/Mbed-TLS/mbedtls/pull/6784#issuecomment-1363721315. Centralizing everything to MD means we can use `MBEDTLS_MD_MAX_SIZE` as it will include PSA-only hashes.
